geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,480 @@
+    final String senderId = (String) args[1];
+    final boolean isCancel = (Boolean) args[2];
+    long maxRate = (Long) args[3];
+    int batchSize = (Integer) args[4];
+
+    final InternalCache cache = (InternalCache) context.getCache();
+
+    if (isCancel) {
+      return cancelWanCopyRegion(context, regionName, senderId);
+    }
+
+    final Region region = cache.getRegion(regionName);
+    if (region == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__REGION__NOT__FOUND, regionName));
+    }
+
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (sender == null) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__NOT__FOUND, senderId));
+    }
+
+    if (!region.getAttributes().getGatewaySenderIds().contains(sender.getId())) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__REGION__NOT__USING_SENDER, regionName,
+              senderId));
+    }
+
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__NOT__RUNNING, senderId));
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__SENDER__SERIAL__AND__NOT__PRIMARY,
+              senderId));
+    }
+
+    try {
+      return executeWanCopyRegionFunctionInNewThread(context, region, regionName, sender, maxRate,
+          batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.WAN_COPY_REGION__MSG__EXECUTION__CANCELED);
+    } catch (ExecutionException e) {
+      Writer buffer = new StringWriter();
+      PrintWriter pw = new PrintWriter(buffer);
+      e.printStackTrace(pw);
+      logger.error("Exception when running wan-copy region command: {}", buffer.toString());
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__EXECUTION__FAILED, e.getMessage()));
+    }
+  }
+
+  private CliFunctionResult executeWanCopyRegionFunctionInNewThread(
+      FunctionContext<Object[]> context,
+      Region region, String regionName, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    ExecutorService executor = LoggingExecutors
+        .newSingleThreadExecutor(getFunctionThreadName(regionName,
+            sender.getId()), true);
```

Comment:
Is this Executor providing the same thread that you later look up by name with the following?
```
for (Thread t : Thread.getAllStackTraces().keySet()) {
      if (t.getName().startsWith(threadBaseName)) {
        found = true;
        t.interrupt();
```
This is a really non-standard, unusual way to do any threading in Java. You should try to get it as close to standard use of threading as possible

Can you make `ExecutorService executor` a final field in this Function class and then just later invoke `executor.shutdownNow()`? `shutdownNow()` will interrupt any thread.

If you need ability to replace the ExecutorService at runtime in a thread-safe way then make use AtomicReference:
```
private final AtomicReference<ExecutorService> executor;
```
Then you can always do:
```
executor.set(LoggingExecutors.newSingleThreadExecutor(getFunctionThreadName(regionName, sender.getId()), true));
``` 
PS: "Java Concurrency In Practice" by Brian Goetz is the gold standard for Java threading

Reply:
@kirklund Thanks for the advice. The problem here is that you could have several instances of this function running and you may want to stop one of them. How could you access the specific instance from a command in order to call the shutdown() method of the `ExecutorService`?
In order to use a more usual way to cancel the threads, I have added static fields to the function class with the `ExecutorService` to be used by any instance of this function and the current executions (`Futures`) so that, if one of them is required to be canceled, you can identify the execution by the senderId and region, get the Future and then invoke cancel on it.
You could also cancel all executions by invoking shutdownNow() on the executor as you suggested. Nevertheless, if you do that, you would not be able to create new executions.

I do not like using static variables but I have not found any better way.